### PR TITLE
Initial prototype gor Gradle resolver

### PR DESCRIPTION
**Pull request by lordofthejars**

### CODE REVIEW

File: gradle/impl-gradle/src/test/java/org/jboss/shrinkwrap/resolver/impl/gradle/ShrinkWrapGradleTestCase.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+public class ShrinkWrapGradleTestCase {
+
+   @Test
+   public void should_return_dependencies_by_scope_in_simple_build_script() {
+      final List<? extends Archive> archives = Gradle.resolver().forProjectDirectory("src/test/resources/simple")
+              .importCompileAndRuntime()
+              .resolve().asList(JavaArchive.class);
+
+      assertThat(archives).extracting("name").contains("slf4j-simple-1.7.5.jar", "slf4j-api-1.7.5.jar");
```
Comment (bartoszmajsak) [inline gradle/impl-gradle/src/test/java/org/jboss/shrinkwrap/resolver/impl/gradle/ShrinkWrapGradleTestCase.java]: `extracting("name")` -> `extracting(Archive::getName)`

File: pom.xml [hunk 2]
```diff
@@ -126,6 +128,11 @@
                 <artifactId>fest-assert</artifactId>
                 <version>${version.fest-assert}</version>
             </dependency>
+            <dependency>
+                <groupId>org.assertj</groupId>
+                <artifactId>assertj-core</artifactId>
```
Comment (bartoszmajsak) [inline pom.xml]: `fest-assert` - this is obsolete, assertj is a successor. They also provide automatic conversion from fest to assertj. If we introduce assertj it's a good time to clean it up

File: gradle/impl-gradle/src/test/java/org/jboss/shrinkwrap/resolver/impl/gradle/ShrinkWrapGradleTestCase.java (hunk 1 above)
Comment (lordofthejars) [inline gradle/impl-gradle/src/test/java/org/jboss/shrinkwrap/resolver/impl/gradle/ShrinkWrapGradleTestCase.java]: Compilation level is Java 7

File: pom.xml (hunk 2 above)
Comment (lordofthejars) [inline pom.xml]: The problem is that I don't know how many tests we should refactor. I think this should be done in another PR.

**Comment by lordofthejars** [comment]
Why Travis is not resolving artifact from jcenter if I have defined in pom? https://github.com/shrinkwrap/resolver/pull/110/files#diff-685e3aa9f29679798de5f4247a92083cR68 Maybe I need to add in parent pom?

File: pom.xml (hunk 2 above)
Comment (bartoszmajsak) [inline pom.xml]: > The problem is that I don't know how many tests we should refactor. 

This is searchable :)

> I think this should be done in another PR.

I believe it should belong in this one. It's really easy.

http://joel-costigliola.github.io/assertj/assertj-core-migrating-from-fest.html

Comment (lordofthejars) [inline pom.xml]: I prefer to change it in another PR, so it is more trackable. The PR is about adding Gradle support not to refactor tests of whole project.

Comment (bartoszmajsak) [inline pom.xml]: But you have introduced new dependency which makes it confusing as you have two very similar libraries for the same thing. Either boyscoutt it or don't use `assertj`.

Comment (MatousJobanek) [inline pom.xml]: Personally, I would leave it for a separate PR. As Alex has mentioned, the target of this PR is implementing Gradle support, not refactoring the current tests...

Comment (bartoszmajsak) [inline pom.xml]: Then I would recommend removing `assertj` entirely and make a new PR introducing it. Otherwise this duality might stay like that forever.

File: gradle/impl-gradle/pom.xml [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    <repositories>
+        <repository>
+            <snapshots>
+                <enabled>false</enabled>
+            </snapshots>
+            <id>central</id>
+            <name>bintray</name>
+            <url>http://jcenter.bintray.com</url>
+        </repository>
+    </repositories>
```
Comment (MatousJobanek) [inline gradle/impl-gradle/pom.xml]: why do we need this repository?

File: gradle/impl-gradle/src/test/java/org/jboss/shrinkwrap/resolver/impl/gradle/ShrinkWrapGradleTestCase.java (hunk 1 above)
Comment (bartoszmajsak) [inline gradle/impl-gradle/src/test/java/org/jboss/shrinkwrap/resolver/impl/gradle/ShrinkWrapGradleTestCase.java]: didn't notice :)

File: pom.xml (hunk 2 above)
Comment (lordofthejars) [inline pom.xml]: I am going to do it just after this is merge so no worries about this

File: gradle/impl-gradle/src/test/java/org/jboss/shrinkwrap/resolver/impl/gradle/ShrinkWrapGradleTestCase.java (hunk 1 above)
Comment (lordofthejars) [inline gradle/impl-gradle/src/test/java/org/jboss/shrinkwrap/resolver/impl/gradle/ShrinkWrapGradleTestCase.java]: I would say users of ShrinkWrap. ShrinkWrap core depends on 7

File: pom.xml (hunk 2 above)
Comment (bartoszmajsak) [inline pom.xml]: Ok next time when I hear from you "in next PR" I would assume it's coming right after the merge ;)

File: gradle/impl-gradle/src/main/java/org/jboss/shrinkwrap/resolver/impl/gradle/GradleDependency.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
+package org.jboss.shrinkwrap.resolver.impl.gradle;
+
+public class GradleDependency {
```
Comment (MatousJobanek) [inline gradle/impl-gradle/src/main/java/org/jboss/shrinkwrap/resolver/impl/gradle/GradleDependency.java]: I haven't found any usage of this class

File: gradle/impl-gradle/src/main/java/org/jboss/shrinkwrap/resolver/impl/gradle/BuildEquippedResolveStage.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
+package org.jboss.shrinkwrap.resolver.impl.gradle;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+public class BuildEquippedResolveStage {
```
Comment (MatousJobanek) [inline gradle/impl-gradle/src/main/java/org/jboss/shrinkwrap/resolver/impl/gradle/BuildEquippedResolveStage.java]: why the name of this class starts with "BuildEquipped" ?

File: gradle/impl-gradle/src/main/java/org/jboss/shrinkwrap/resolver/impl/gradle/GradleEffectiveModel.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+import org.gradle.tooling.model.idea.IdeaSingleEntryLibraryDependency;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class GradleEffectiveModel {
```
Comment (MatousJobanek) [inline gradle/impl-gradle/src/main/java/org/jboss/shrinkwrap/resolver/impl/gradle/GradleEffectiveModel.java]: Is this really an effective model or more likely only set of all dependencies?

File: gradle/impl-gradle/pom.xml (hunk 3 above)
Comment (MatousJobanek) [inline gradle/impl-gradle/pom.xml]: ah, I see - because of this dependency: gradle-tooling-api

**Comment by MatousJobanek** [comment]
@lordofthejars as for jcenter - maybe it has problems with the id: central

File: gradle/impl-gradle/pom.xml [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.gradle</groupId>
+            <artifactId>gradle-tooling-api</artifactId>
+            <version>${version.gradle-tooling-api}</version>
```
Comment (MatousJobanek) [inline gradle/impl-gradle/pom.xml]: Shouldn't be the version updated to the latest one? Or are there any incompatibility problems?

File: gradle/impl-gradle/pom.xml (hunk 3 above)
Comment (MatousJobanek) [inline gradle/impl-gradle/pom.xml]: But why the snapshots are enabled and what is the difference between this repo and the jfrog one used in `impl-gradle-embedded-archive` module? https://github.com/shrinkwrap/resolver/blob/master/gradle/impl-gradle-embedded-archive/pom.xml#L73-L77

Comment (lordofthejars) [inline gradle/impl-gradle/pom.xml]: Because gradle-tooling is released there and not maven central.

File: gradle/impl-gradle/src/main/java/org/jboss/shrinkwrap/resolver/impl/gradle/BuildEquippedResolveStage.java (hunk 5 above)
Comment (lordofthejars) [inline gradle/impl-gradle/src/main/java/org/jboss/shrinkwrap/resolver/impl/gradle/BuildEquippedResolveStage.java]: To follow the name convention in Maven part `PomEquippedResolveStrategy`

File: gradle/impl-gradle/src/main/java/org/jboss/shrinkwrap/resolver/impl/gradle/GradleEffectiveModel.java (hunk 6 above)
Comment (lordofthejars) [inline gradle/impl-gradle/src/main/java/org/jboss/shrinkwrap/resolver/impl/gradle/GradleEffectiveModel.java]: It is the resolution of all dependencies

File: gradle/impl-gradle/pom.xml (hunk 7 above)
Comment (lordofthejars) [inline gradle/impl-gradle/pom.xml]: I am using 3 because it is the one defined in parent pom, but we can update to 3.3 if we want.

File: gradle/impl-gradle/src/main/java/org/jboss/shrinkwrap/resolver/impl/gradle/GradleDependency.java (hunk 4 above)
Comment (lordofthejars) [inline gradle/impl-gradle/src/main/java/org/jboss/shrinkwrap/resolver/impl/gradle/GradleDependency.java]: True this class worked when I thought we could resolve individual dependencies. I am going to remove for now.

File: gradle/impl-gradle/pom.xml (hunk 3 above)
Comment (lordofthejars) [inline gradle/impl-gradle/pom.xml]: I will check because probably this is the main problem here and Travis.

File: gradle/impl-gradle/pom.xml (hunk 7 above)
Comment (MatousJobanek) [inline gradle/impl-gradle/pom.xml]: 3.3 is the version what I had expected. I thought that you need/want to use the version for the feature, that's the reason why I wondered why you haven't updated the version.

Comment (lordofthejars) [inline gradle/impl-gradle/pom.xml]: No I read that 3.X is ok, if you want I update.

**Comment by MatousJobanek** [comment]
@mmatloka maybe something interesting for you (as you were the one who has been updating the gradle-tooling-api version)

File: gradle/impl-gradle/src/main/java/org/jboss/shrinkwrap/resolver/impl/gradle/GradleEffectiveDependencies.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Class that groups dependencies by scope.
+ */
+public class GradleEffectiveDependencies {
+
+   private Map<ScopeType, List<File>> effectiveModelGroupedByScope = new HashMap<>();
```
Comment (mmatloka) [inline gradle/impl-gradle/src/main/java/org/jboss/shrinkwrap/resolver/impl/gradle/GradleEffectiveDependencies.java]: @ALRubinger style, finals everywhere ;)

**Comment by MatousJobanek** [comment]
LGTM - good job Alex 👍
